### PR TITLE
Upgrade MathJax support from v2/v3 to v3/v4

### LINKED ISSUE #7598: [FEATURE]: Update library to handle MathJax v4 and remove v2 support
### Description

Update library to handle MathJax v4 and remove v2 support.

### Why should this feature be added?

- v4 was recently released and mostly adds features which could be useful to users
- v2 contains a [security vulnerability](https://github.com/mathjax/MathJax/issues/3074) (ReDoS attack). While this issue doesn't affect the library much (only used as a dev dependency), it shows up as a warning with `npm audit`.
- The [latest v2 release](https://github.com/mathjax/MathJax/releases/tag/2.7.9) came out in 2020
- Removing v2 support would simplify the code

### Notes

- This would be a breaking change
- The [upgrade process](https://docs.mathjax.org/en/latest/upgrading/v3.html) looks pretty straightforward
- Jupyter bundles a version of MathJax and the library needs to remain compatible with that

### PR DESCRIPTION
**Pull request by emilykl**
Update plotly.js to be compatible with MathJax v4, and drop support for MathJax v2.

Closes #7598.

Due to some logic improvements, this should also make MathJax rendering faster.

- **Modify `src/lib/svg_text_utils.js` to use MathJax v3/v4 syntax to render math symbols**
  - MathJax v3 and v4 use the same syntax, so we can get rid of the if/else cases we used to handle v2 and v3
- **Improve performance by not calling `MathJax.startup.defaultReady()` on every individual render**
  - This was _very slow_ and also not recommended usage of the MathJax API
  - Instead, create a secondary MathJax document upfront that we can use for our purposes, without affecting the page's main MathJax document, and call `MathDocument.convert(tex)` rather than `MathJax.typeset(tex)`
    - Side note: This seems to be more or less the approach proposed [here](https://github.com/plotly/plotly.js/issues/4563#issuecomment-1004029245) by @akhmerov in 2022
  - This also means we no longer have to reset the config after rendering, since the config is applied only to our internal MathJax document
  - This approach does involve three places where we reference functions/objects in the `MathJax._.mathjax` namespace. While not officially documented, the pattern is used in several places ([one](https://docs.mathjax.org/en/v4.0/web/retry.htm), [two](https://docs.mathjax.org/en/v4.1/input/asciimath.html), [three](https://docs.mathjax.org/en/v4.1/output/mathml.html)) in the MathJax docs for advanced configuration
    - The MathJax API is mainly intended for rendering entire documents/pages in one shot, so it doesn't directly expose functions for rendering individual expressions one-at-a-time
  - As a nice side effect, this change should make MathJax rendering in plotly much faster
- **Update image baselines**
  - There are visible differences, but no regressions; all the changes are minor typsetting differences in character spacing or line height, which makes sense for a new MathJax version
  - Exception: In the `mathjax3__mathjax` mock, the pie labels which were previously hidden, are now visible as raw MathJax. We don't support rendering MathJax in pie labels. Now at least the behavior is consistent between v3 and v4. However we could change the behavior to hide MathJax text strings in places where we don't support rendering MathJax.
- **Update all testing infrastructure to use MathJax v3 and v4** rather than v2 and v3.
  - **NOTE**: I've done this by changing all v2 references to v4. This means that v4 is now the "default" in the CI, unless v3 is explicitly specified
  - This includes some updates to the MathJax devtools dashboard homepages 

### Steps for testing

- Check out this branch
- Run the test dashboard (`npm start`) **which uses MathJax 4**
- Verify that every mock with `mathjax` in the name (and the other baselines changed here) looks OK and matches the new baselines on this branch
  - Note: The `mathjax` mock itself contains several strings of TeX which don't render due to us not supporting MathJax in pie labels; this is not new, and it's (somewhat) deliberate, to lock in the fallback behavior
- Visit the **MathJax 3** test dashboard (`http://localhost:3000/devtools/test_dashboard/index-mathjax3.html`) and verify again that everything looks fine

### CODE REVIEW

File: src/lib/svg_text_utils.js [hunk 1]
```diff
@@ -90,6 +94,11 @@ exports.convertToTspans = function(_context, gd, _callback) {
                     return;
                 }
 
+                // Now that the MathJax render has finished, re-hide the source text.
+                // We hid it earlier, too, but since this callback runs async,
+                // another function may have made it visible again
```
Comment (emilykl) [inline src/lib/svg_text_utils.js:99]: add a TODO to investigate this further

File: src/lib/svg_text_utils.js [hunk 2]
```diff
@@ -185,154 +192,87 @@ function cleanEscapesForTex(s) {
...
-        var tmpNode = tmpDiv.node();
+            });
 
-        return MathJaxVersion === 2 ?
-            MathJax.Hub.Typeset(tmpNode) :
-            MathJax.typeset([tmpNode]);
+        // Strip $…$ delimiters and clean up escape charaters,
+        // then pass the result to MathDocument.convert() to get an SVG element back
+        const texMath = cleanEscapesForTex(_texString).replace(/^\$+|\$+$/g, '');
+        return MathJax._.mathjax.mathjax.handleRetriesFor(function() {
```
Comment (camdecoster) [inline src/lib/svg_text_utils.js:253]: Why is this call to `handleRetriesFor` required?

File: src/lib/svg_text_utils.js [hunk 3]
```diff
@@ -185,154 +192,87 @@ function cleanEscapesForTex(s) {
...
-        setRenderer3();
-        MathJax.startup.defaultReady();
-
-        MathJax.startup.promise.then(function() {
-            initiateMathJax();
-            finalizeMathJax();
-
-            resetRenderer3();
-            resetConfig3();
+    // Initialize, render MathJax, then call _callback()
```
Comment (camdecoster) [inline src/lib/svg_text_utils.js]:
```suggestion
```

File: src/lib/svg_text_utils.js [hunk 4]
```diff
@@ -185,154 +192,87 @@ function cleanEscapesForTex(s) {
...
-        }
-    };
-
-    var setRenderer3 = function() {
-        originalRenderer = MathJax.config.startup.output;
-        if(originalRenderer !== 'svg') {
-            MathJax.config.startup.output = 'svg';
+    var tmpDiv;
+
+    const initiateMathJax = function() {
```
Comment (camdecoster) [inline src/lib/svg_text_utils.js:222]: Should this be an async function?

File: draftlogs/7898_change.md [hunk 5]
```diff
@@ -0,0 +1 @@
+- Add support for MathJax v4, and **drop support for v2** [[#7898](https://github.com/plotly/plotly.js/pull/7898)]
```
Comment (camdecoster) [inline draftlogs/7898_change.md]:
```suggestion
- **Breaking:** Add support for MathJax v4, and **drop support for v2** [[#7898](https://github.com/plotly/plotly.js/pull/7898)]
```

File: test/image/baselines/legend_mathjax_title_and_items.png
Comment (camdecoster) [inline test/image/baselines/legend_mathjax_title_and_items.png:1]: Do you now why the legend got smaller?

File: test/image/baselines/mathjax3___mathjax.png
Comment (camdecoster) [inline test/image/baselines/mathjax3___mathjax.png:1]: Is the MathJax specifier supposed to be showing up in the pie chart? Is this the fallback behavior?

File: src/lib/svg_text_utils.js (hunk 2 above)
Comment (emilykl) [reply src/lib/svg_text_utils.js:253]: Found a page in the docs which explains it: https://docs.mathjax.org/en/v4.0/web/retry.html

From what I understand, it sounds like MathJax dynamically loads some features/extensions only as needed, and the internal synchronous functions are designed to error out if a required feature is not available, with a wrapper function catching the failure, loading the required feature, then retrying the function. The wrapper function returns a Promise which resolves when the wrapped function eventually completes successfully.

`handleRetriesFor()` is that wrapper function.

`.convert()` is a synchronous function, so wrapping it inside `handleRetriesFor()` has the additional impact of turning the call to `.convert()` into an asynchronous function.

File: src/lib/svg_text_utils.js (hunk 4 above)
Comment (emilykl) [reply src/lib/svg_text_utils.js:222]: I believe since `initiateMathJax()` always returns a Promise, there's no programmatic reason to make it an async function (all that would achieve is making the function return a Promise, but it's doing that anyway).

I guess it wouldn't cause any problems to add `async` to the function definition, but I don't think we've done that anywhere else in the codebase.

File: test/image/baselines/legend_mathjax_title_and_items.png
Comment (emilykl) [reply test/image/baselines/legend_mathjax_title_and_items.png:1]: I don't know, but I did add a new mock [`mathjax-font-size`](https://github.com/plotly/plotly.js/pull/7898/changes#diff-0c54e6eb5a778c9c4dc3e20f960912e3da75146a668fffac1433bd97eeffc146) to ensure that the rendered MathJax is picking up the font size of the original text.

Since that seems to be working, I think this change might just be a styling difference between MathJax v2 and v4.

File: test/image/baselines/mathjax3___mathjax.png
Comment (emilykl) [reply test/image/baselines/mathjax3___mathjax.png:1]: @camdecoster I think I said something different in our call, but the actual reason why the raw MathJax string is showing up here because the library doesn't yet render MathJax for pie labels. So the entire rendering step is skipped and the string is shown as-is. (The exact same string is rendered 'correctly' in the legend to the right.)

The raw MathJax string shown here is present since the `mathjax` baseline image was first added to the repo. The fact that the string _wasn't_ shown in the `mathjax3__` version of the mock is the outlier. I'm pretty neutral on which behavior is more correct.

File: src/lib/svg_text_utils.js [hunk 6]
```diff
@@ -114,7 +125,11 @@ exports.convertToTspans = function(_context, gd, _callback) {
                     height: h0,
                     preserveAspectRatio: 'xMinYMin meet'
                 })
-                .style({overflow: 'visible', 'pointer-events': 'none'});
+                .style({
+                    overflow: 'visible',
+                    'pointer-events': 'none',
+                    'font-size': config.fontSize + 'px',
```
Comment (emilykl) [inline src/lib/svg_text_utils.js:131]: Note: this addition fixes a bug where MathJax expressions would not render larger than font size 16 in plotly

File: test/image/baselines/mathjax3___mathjax.png
Comment (archmoj) [reply test/image/baselines/mathjax3___mathjax.png:1]: The MathJax specifier supposed to be showing up in the pie chart text (similar to the MathJax v2 baseline).
Please note that `.attr('data-notex', 1)` is set in `src/traces/pie/plot.js` as well as few other places in the codebase. It means that MathJax is currently not supported there.

File: package.json [hunk 7]
```diff
@@ -115,8 +116,8 @@
   },
   "devDependencies": {
     "@biomejs/biome": "2.2.0",
-    "@plotly/mathjax-v2": "npm:mathjax@2.7.5",
-    "@plotly/mathjax-v3": "npm:mathjax@^3.2.2",
+    "@plotly/mathjax-v3": "npm:mathjax@3.2.2",
+    "@plotly/mathjax-v4": "npm:mathjax@^4.1.2",
```
Comment (archmoj) [inline package.json]: Here the caret (^) sign is removed for MathJax v3 pinning down the version of dependency. IMHO this is a good practice to avoid unwanted test breakages as a result of future minor/patch releases of MathJax v3.
If that's the case, then we should possibly do the same for MathJax v4 i.e. 
```json
    "@plotly/mathjax-v4": "npm:mathjax@4.1.3",
```
Please note that the latest v4 is now at `4.1.3`:
https://www.npmjs.com/package/mathjax?activeTab=versions

File: src/lib/svg_text_utils.js (hunk 4 above)
Comment (camdecoster) [reply src/lib/svg_text_utils.js:222]: That's fair. I guess the signal is that this function is used in the promise chain down below.

File: src/lib/svg_text_utils.js (hunk 2 above)
Comment (camdecoster) [reply src/lib/svg_text_utils.js:253]: Thanks for the explanation!

File: draftlogs/7898_change.md (hunk 5 above)
Comment (archmoj) [inline draftlogs/7898_change.md]: You may consider adding multiple draftlogs for this PR e.g. 
`7898_add`: Add mathjax v4 support
`7898_remove`: Remove mathjax v2 support
`7898_fix`: Improve mathjax v3 rendering 
That way the changes would clearly be reflected in the plotly.js v4 changelog for users as well as QA.

File: test/image/baselines/mathjax3___mathjax.png
Comment (camdecoster) [reply test/image/baselines/mathjax3___mathjax.png:1]: Got it. It's good that's it showing in v4.

File: test/image/baselines/legend_mathjax_title_and_items.png
Comment (camdecoster) [reply test/image/baselines/legend_mathjax_title_and_items.png:1]: It's not a blocker. If something needs to be fixed, we can do it post v4 release.

File: package.json (hunk 7 above)
Comment (camdecoster) [reply package.json]: Did you have trouble with that in the past? If not, it seems fine to leave this unbounded and rely on the lock file to keep us in sync.

**Review by camdecoster — APPROVED** [review]
It's working for me. Good work!

File: package.json (hunk 7 above)
Comment (archmoj) [reply package.json]: Yes I recall noticing baseline changes between MathJax v2 patch and minor releases.
If you decide to unpin `mathjax-v4` then `mathjax-v3` may also be reverted to the unpin version i.e. 
```json 
    "@plotly/mathjax-v3": "npm:mathjax@^3.2.2",
```

File: draftlogs/7898_change.md (hunk 5 above)
Comment (emilykl) [reply draftlogs/7898_change.md]: Done

File: package.json (hunk 7 above)
Comment (emilykl) [reply package.json]: Discussed with @camdecoster, we will unpin both v3 and v4 for now and see how it goes.

If we experience too many CI failures as a result, we will go back to pinning exact versions.

**Comment by emilykl** [comment]
Note: While QAing these changes in Jupyter Notebook (v7), I discovered that the latest Plotly release (plotly.py 6.9.0, plotly.js 3.7.0) does not render MathJax equations in Jupyter.

There is an open plotly.py issue https://github.com/plotly/plotly.py/issues/1740 capturing the problem for JupyterLab; I am not sure how long the issue has been present in Jupyter Notebook as well, but possibly since the release of Jupyter Notebook 7. It seems at some point, Jupyter stopped exposing a fully-loaded global `MathJax` variable, so we're not able to hook into the existing MathJax script loaded on the page.

All this to say, the changes in this PR _also_ don't work in JupyterLab / Jupyter Notebook; however, since the issue is previously existing, this isn't a regression. We will move ahead with these changes while also investigating how to restore MathJax rendering in Plotly charts in Jupyter.

File: test/image/make_baseline.py [hunk 8]
```diff
@@ -46,6 +44,11 @@
 
     mathjax_version = 3
     print("Kaleido using MathJax v3")
+else:
+    mathjax_version = 4
+    # Kaleido still defaults to MathJax v2, so we need to explicitly specify the path to MathJax v4
+    mathjax = "https://cdn.jsdelivr.net/npm/mathjax@4.1.2/tex-svg.js"
```
Comment (archmoj) [inline test/image/make_baseline.py]:
```suggestion
    mathjax = "https://cdn.jsdelivr.net/npm/mathjax@4.1.3/tex-svg.js"
```

File: test/image/make_exports.py [hunk 9]
```diff
@@ -8,7 +8,8 @@
 dirOut = os.path.join(root, "build", "test_images")
 
 pio.templates.default = "none"
-pio.kaleido.scope.plotlyjs = os.path.join(root, "build", "plotly.js")
+pio.defaults.plotlyjs = os.path.join(root, "build", "plotly.js")
+pio.defaults.mathjax = "https://cdn.jsdelivr.net/npm/mathjax@4.1.2/tex-svg.js"
```
Comment (archmoj) [inline test/image/make_exports.py]:
```suggestion
pio.defaults.mathjax = "https://cdn.jsdelivr.net/npm/mathjax@4.1.3/tex-svg.js"
```

File: README.md [hunk 10]
```diff
@@ -95,17 +95,23 @@ While non-minified source files may contain characters outside UTF-8, it is reco
...
 ### MathJax
-You can load either version two or version three of MathJax files. For example:
+plotly.js supports using MathJax v3 or v4 to render mathematical expressions within plots.
+
+MathJax is not included in the plotly.js bundle, so it must be loaded separately as a script on the webpage.
+
+For MathJax v4:
 ```html
-<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js?config=TeX-AMS-MML_SVG.js"></script>
+<script src="https://cdn.jsdelivr.net/npm/mathjax@4.1.2/tex-svg.js"></script>
```
Comment (archmoj) [inline README.md]:
```suggestion
<script src="https://cdn.jsdelivr.net/npm/mathjax@4.1.3/tex-svg.js"></script>
```

File: src/lib/svg_text_utils.js (hunk 6 above)
Comment (archmoj) [reply src/lib/svg_text_utils.js:131]: Excellent. Thank you.

**Comment by archmoj** [comment]
Thanks @emilykl for the great PR.
It looks like a new baseline titled `mathjax3___mathjax-font-size.png` is missing in respect to the new mock?

**Comment by emilykl** [comment]
> Thanks @emilykl for the great PR. It looks like a new baseline titled `mathjax3___mathjax-font-size.png` is missing in respect to the new mock?

Good catch, I forgot new mocks had to be manually added to the `mathjax3` list. I will fix.
